File: javascript/specs/elements/field_with_items.spec.js
import { parse } from '../..';
import { Item } from '../../lib/elements/item.js';

describe('Field', () => {
    let field;
    
    beforeEach(() => {
        const document = parse(`
            field:
            - item
            - other
            `.trim());
            
            field = document.field('field');
        });
        
        it('is untouched after initialization', () => {
            expect(field._touched).toBeUndefined();
        });
        
        it('has only untouched items after initialization', () => {
            for (const item of field.items()) {
                expect(item._touched).toBeUndefined();
            }
        });
        
        describe('items()', () => {
            let items;
            
            beforeEach(() => {
                items = field.items();
            });
            
            it('touches the field itself', () => {
                expect(field._touched).toBe(true);
            });
            
            it('does not touch the field items', () => {
                for (const item of items) {
                    expect(item._touched).toBeUndefined();
                }
            });
            
            it('returns the items', () => {
                for (const item of items) {
                    expect(item).toBeInstanceOf(Item);
                }
            });
        });
        
        describe('optionalStringValues()', () => {
            let values;
            
            beforeEach(() => {
                values = field.optionalStringValues();
            });
            
            it('returns the values', () => {
                expect(values).toEqual(['item', 'other']);
            });
            
            it('touches the field itself', () => {
                expect(field._touched).toBe(true);
            });
            
            it('touches all field items', () => {
                for (const item of field.items()) {
                    expect(item._touched).toBe(true);
                }
            });
        });
        
        describe('requiredValues(loader)', () => {
            let values;
            
            beforeEach(() => {
                values = field.requiredValues(value => value.toUpperCase());
            });
            
            it('returns the processed values', () => {
                expect(values).toEqual(['ITEM', 'OTHER']);
            });
            
            it('touches the element', () => {
                expect(field._touched).toBe(true);
            });
            
            it('touches all field items', () => {
                for (const item of field.items()) {
                    expect(item._touched).toBe(true);
                }
            });
        });
        
        describe('length()', () => {
            it('returns the number of items', () => {
                expect(field.length()).toBe(2);
            });
        });
        
        describe('toString()', () => {
            it('returns a debug abstraction', () => {
                expect(field.toString()).toEqual('[object Field key=field items=2]');
            });
        });
        
        describe('toStringTag symbol', () => {
            it('returns a custom tag', () => {
                expect(Object.prototype.toString.call(field)).toEqual('[object Field]');
            });
        });
        
        describe('touch()', () => {
            beforeEach(() => {
                field.touch();
            });
            
            it('touches the field itself', () => {
                expect(field._touched).toBe(true);
            });
            
            it('touches the field items', () => {
                for (const item of field.items()) {
                    expect(item._touched).toBe(true);
                }
            });
        });
    });

File: ruby/spec/elements/field_with_items.spec.rb
# frozen_string_literal: true

input = <<~DOC.strip
field_with_items:
- item
- other
DOC

describe Enolib::Field do
  before(:each) do
    @field = Enolib.parse(input).field('field_with_items')
  end

  it 'is untouched after initialization' do
    expect(@field.instance_variable_defined?(:@touched)).to be false
  end

  it 'has only untouched items after initialization' do
    @field.items.each do |item|
      expect(item.instance_variable_defined?(:@touched)).to be false
    end
  end

  describe '#required_string_values' do
    before(:each) do
      @items = @field.required_string_values
    end

    it 'returns the values' do
      expect(@items).to eq(['item', 'other'])
    end

    it 'touches the field itself' do
      expect(@field.instance_variable_defined?(:@touched)).to be true
    end

    it 'touches all field items' do
      @field.items.each do |item|
        expect(item.instance_variable_defined?(:@touched)).to be true
      end
    end
  end

  describe '#required_values' do
    context 'with a loader block' do
      before(:each) do
        @items = @field.required_values { |value| value.upcase }
      end

      it 'applies the loader' do
        expect(@items).to eq(['ITEM', 'OTHER'])
      end

      it 'touches the field' do
        expect(@field.instance_variable_defined?(:@touched)).to be true
      end

      it 'touches all field items' do
        @field.items.each do |item|
          expect(item.instance_variable_defined?(:@touched)).to be true
        end
      end
    end

    context 'with a loader Proc' do
      before(:each) do
        @items = @field.required_values(proc { |value| value.upcase })
      end

      it 'applies the loader' do
        expect(@items).to eq(['ITEM', 'OTHER'])
      end

      it 'touches the field' do
        expect(@field.instance_variable_defined?(:@touched)).to be true
      end

      it 'touches all field items' do
        @field.items.each do |item|
          expect(item.instance_variable_defined?(:@touched)).to be true
        end
      end
    end
  end

  describe '#length' do
    it 'returns the number of items in the field' do
      expect(@field.length).to be(2)
    end
  end

  describe '#touch' do
    before(:each) do
      @field.touch
    end

    it 'touches the field itself' do
      expect(@field.instance_variable_defined?(:@touched)).to be true
    end

    it 'touches the field items' do
      @field.items.each do |item|
        expect(item.instance_variable_defined?(:@touched)).to be true
      end
    end
  end

  describe '#to_s' do
    it 'returns a debug abstraction' do
      expect(@field.to_s).to eq('#<Enolib::Field key=field_with_items items=2>')
    end
  end
end
